### PR TITLE
KOGITO-2207 Introduce ProcessInstanceReadMode to allow read only proc…

### PR DESCRIPTION
**Pull request by cristianonicolai**
…ess instances

Assembly: 
- https://github.com/kiegroup/kogito-examples/pull/309
- https://github.com/kiegroup/kogito-runtimes/pull/652
- https://github.com/kiegroup/kie-docs/pull/2746

### CODE REVIEW

**Comment by cristianonicolai** [comment]
jenkins please retest this

**Comment by cristianonicolai** [comment]
jenkins please retest this

**Comment by cristianonicolai** [comment]
@tiagodolphine exactly, take a look in the docs part, I tried to explain a bit more in details there, see https://github.com/kiegroup/kie-docs/pull/2746/files#diff-551ebcaf10443155f86f0ebb04bda41cR10

File: addons/persistence/filesystem-persistence-addon/src/main/java/org/kie/kogito/persistence/filesystem/FileSystemProcessInstances.java [hunk 1]
```diff
@@ -64,24 +69,40 @@ public FileSystemProcessInstances(Process<?> process, Path storage, ProcessInsta
     }
 
     @Override
-    public Optional findById(String id) {
+    public Integer size() {
+        try (Stream<Path> stream = Files.walk(storage)) {
+            Long count = stream.filter(file -> !Files.isDirectory(file)).count();
+            return count.intValue();
+        } catch (IOException e) {
+            throw new RuntimeException("Unable to read process instances ", e);
```
[the remark this answers is not in the document]
Comment (cristianonicolai) [reply addons/persistence/filesystem-persistence-addon/src/main/java/org/kie/kogito/persistence/filesystem/FileSystemProcessInstances.java]: oops :)

File: addons/persistence/filesystem-persistence-addon/src/test/java/org/kie/persistence/filesystem/FileSystemProcessInstancesTest.java [hunk 2]
```diff
@@ -46,12 +49,46 @@
...
+        process.setProcessInstancesFactory(new FileSystemProcessInstancesFactory());
+        process.configure();
+
+        ProcessInstance<BpmnVariables> processInstance = process.createInstance(BpmnVariables.create(Collections.singletonMap("test", "test")));
+
+        processInstance.start();
+
+        ProcessInstances<BpmnVariables> instances = process.instances();
+        assertThat(instances.size()).isOne();
+        assertThatNullPointerException().isThrownBy(() -> instances.findById(processInstance.id(), ProcessInstanceReadMode.READ_ONLY).get().abort());
```
[the remark this answers is not in the document]
Comment (cristianonicolai) [reply addons/persistence/filesystem-persistence-addon/src/test/java/org/kie/persistence/filesystem/FileSystemProcessInstancesTest.java]: managed to change so now calling any operation on a read only process instance will throw a `UnsupportedOperationException`.

File: addons/persistence/filesystem-persistence-addon/src/test/java/org/kie/persistence/filesystem/FileSystemProcessInstancesTest.java [hunk 3]
```diff
@@ -46,12 +49,46 @@
...
+        instances.findById(processInstance.id()).get().abort();
+        assertThat(instances.size()).isZero();
+    }
+
+    @Test
+    void testValuesReadMode() {
```
[the remark this answers is not in the document]
Comment (cristianonicolai) [reply addons/persistence/filesystem-persistence-addon/src/test/java/org/kie/persistence/filesystem/FileSystemProcessInstancesTest.java:113]: extended the test case to cover that as well. Thanks

**Comment by tiagodolphine** [comment]
Ok, so I got the point... one other thing it is worth and it would be important to scale at a huge number of process instances, is to get rid of the `values()` method https://github.com/kiegroup/kogito-runtimes/blob/a26c98a4b251c3cce513cd0310691288a291c7e1/api/kogito-api/src/main/java/org/kie/kogito/process/ProcessInstances.java#L25 because it brings all instances into memory + all the marshalling process... I mean this does not scale, and on the REST APIs perspective, the getting all instances does not make sense we have data-index to fetch the instances/filter/pagination etc... WDYT? I'm wondering why this method is really needed and if it would be feasible to remove it.

**Comment by cristianonicolai** [comment]
jenkins please retest this

**Comment by cristianonicolai** [comment]
@tiagodolphine I see your point and I agree that for any querying, the data index should be used, I just feel that we would restrict too much, even for demos? that the process id would need to be stored always, as we would only provide methods based on the process id.  Iwould say that use is disincouraged but not sure we should remove it entirely.

**Comment by cristianonicolai** [comment]
jenkins please retest this

**Comment by cristianonicolai** [comment]
jenkins please retest this

**Comment by cristianonicolai** [comment]
jenkins please retest this

**Pull request merged**